src/twisted/web/http.py

```diff
@@ -861,30 +861,56 @@ def requestReceived(self, command, path, version):
         # Argument processing
         args = self.args
         ctype = self.requestHeaders.getRawHeaders(b'content-type')
+        clength = self.requestHeaders.getRawHeaders(b'content-length')
         if ctype is not None:
             ctype = ctype[0]
 
-        if self.method == b"POST" and ctype:
+        if clength is not None:
+            clength = clength[0]
+
+        if self.method == b"POST" and ctype and clength:
             mfd = b'multipart/form-data'
             key, pdict = _parseHeader(ctype)
+            pdict["CONTENT-LENGTH"] = clength
             if key == b'application/x-www-form-urlencoded':
                 args.update(parse_qs(self.content.read(), 1))
             elif key == mfd:
                 try:
-                    cgiArgs = cgi.parse_multipart(self.content, pdict)
+                    if _PY37PLUS:
+                        cgiArgs = cgi.parse_multipart(
+                            self.content, pdict, encoding='utf8',
+                            errors="surrogateescape")
+                    else:
+                        cgiArgs = cgi.parse_multipart(self.content, pdict)
 
-                    if _PY3:
-                        # parse_multipart on Python 3 decodes the header bytes
-                        # as iso-8859-1 and returns a str key -- we want bytes
-                        # so encode it back
+                    if not _PY37PLUS and _PY3:
+                        # The parse_multipart function on Python 3
+                        # decodes the header bytes as iso-8859-1 and
+                        # returns a str key -- we want bytes so encode
+                        # it back
                         self.args.update({x.encode('iso-8859-1'): y
                                           for x, y in cgiArgs.items()})
+                    elif _PY37PLUS:
+                        # The parse_multipart function on Python 3.7+
+                        # decodes the header bytes as iso-8859-1 and
+                        # decodes the body bytes as utf8 with
+                        # surrogateescape -- we want bytes
+                        self.args.update({
+                            x.encode('iso-8859-1'): \
+                            [z.encode('utf8', "surrogateescape") for z in y]
```

Comment:
Body components that represent files are still returned as bytes.  You can confirm this with the following unpleasant snippet:

```python
import io
import pprint
import cgi
from http import server
from http import HTTPStatus
import socketserver


PORT = 8001

class Handler(server.BaseHTTPRequestHandler):
    content = b'''
<html>
<form action="/" method="post" enctype="multipart/form-data">
What files are you sending? <INPUT type="file" name="files"><BR>
<input type="submit">
</form>
'''

    def do_GET(self):
        self.send_response(HTTPStatus.OK)
        self.send_header("Content-type", "text/html")
        self.send_header("Content-length", str(len(self.content)))
        self.end_headers()
        self.wfile.write(self.content)

    def do_POST(self):
        _, pdict = cgi.parse_header(self.headers['Content-type'])
        pdict['CONTENT-LENGTH'] = self.headers['Content-length']
        pdict = {key: value.encode('charmap') for key, value in pdict.items()}
        parsed = cgi.parse_multipart(self.rfile, pdict)
        bio = io.StringIO()
        pprint.pprint(parsed, stream=bio)
        self.wfile.write(bio.getvalue().encode('utf-8'))


with socketserver.TCPServer(("", PORT), Handler) as httpd:
    print("serving at port", PORT)
    httpd.serve_forever()
```

That means you'll need to do something like `[z.encode('utf-8', 'surrogate-escape') if isinstance(z, str) else z for z in y]`.

Reply:
sob let me fix it